### ISSUE TITLE
csolution skips silently incorrect condition elements

### ISSUE THREAD

**Issue by fred-r** [Open-CMSIS-Pack/devtools#744]
Hi,

I have a DFP pack for the STM32U5 series.

I have a component defined like this in the DPF pack:
```
    <condition id="STM32U575">
      <require Dvendor="STMicroelectronics:13" Dnameazerty="STM32U575*"/>
    </condition>

    <condition id="STM32U585">
      <require Dvendor="STMicroelectronics:13" Dname="STM32U585*"/>
    </condition>

...

        <!-- startup files -->
        <file category="source" condition="STM32U575" name="Source/startup_stm32u575xx.c" version="0.0.1"/>
        <file category="source" condition="STM32U585" name="Source/startup_stm32u585xx.c" version="0.0.1"/>
...
```

I have a project defined like this:
```
csolution:
solution:
  target-types:
    - type: B-U585I-IOT02A
      device: STM32U585AIIxQ
....
cproject:
...
    - component: STMicroelectronics::Device:Startup
...
```

Then I invoke:
```
csolution.exe -s hal_gpio_toggle.csolution.yml convert
```

And obviously the incorrect condition member __Dnameazerty="STM32U575*"__ is just skipped, and the condition __STM32U575__ evaluates to true (which is incorrect).

```
$ cd . /RTE/Device
$ cd STM32U585AIIxQ/
$ ls -al
-rw-r--r-- 1 frq09468 1049089  1358 Dec 22 17:01  stm32u575xx_flash.icf
-r--r--r-- 1 frq09468 1049089  1358 Dec 22 17:01 'stm32u575xx_flash.icf.base@0.0.1'
-rw-r--r-- 1 frq09468 1049089  5109 Dec 22 17:01  STM32U575xx_FLASH.ld
-r--r--r-- 1 frq09468 1049089  5109 Dec 22 17:01 'STM32U575xx_FLASH.ld.base@0.0.1'
-rwxr-xr-x 1 frq09468 1049089   860 Dec 22 17:01  stm32u575xx_flash.sct*
-r-xr-xr-x 1 frq09468 1049089   860 Dec 22 17:01 'stm32u575xx_flash.sct.base@0.0.1'*
-rw-r--r-- 1 frq09468 1049089  1317 Dec 22 17:01  stm32u585xx_flash.icf
-r--r--r-- 1 frq09468 1049089  1317 Dec 22 17:01 'stm32u585xx_flash.icf.base@0.0.1'
-rw-r--r-- 1 frq09468 1049089  5112 Dec 22 17:01  STM32U585xx_FLASH.ld
-r--r--r-- 1 frq09468 1049089  5112 Dec 22 17:01 'STM32U585xx_FLASH.ld.base@0.0.1'
-rwxr-xr-x 1 frq09468 1049089   861 Dec 22 17:01  stm32u585xx_flash.sct*
-r-xr-xr-x 1 frq09468 1049089   861 Dec 22 17:01 'stm32u585xx_flash.sct.base@0.0.1'*
```

**Comment by fred-r** [comment]
I would expect csolution to report a failure in the condition and the impossibility to evaluate it, or at least a warning.

**Comment by fred-r** [comment]
@vvestm @abostm

**Comment by jkrech** [comment]
@fred-r  I would have expected you to run a schema check as well as packchk on your DFP before using it. 
The tool is ignoring any attribute that it does not understand as it maybe a future extension.
Since the only remaining attribute is the Dvendor the condition rightfully resolves to true.

**Comment by ReinhardKeil** [comment]
Did you use packchk to verify the pack?  Does this detect the mistake in the conditions?

**Comment by fred-r** [comment]
Interesting point.
As I ma in "development mode", I use "local packs" to avoid creating a .pack and installing it for each change.
But, your point about tolerating future extensions make sense.
Why not raising at least a warning if not failing ?

Even if it is a "future extension", it means that this conditions has not been evaluated as you may expect because the tool is not in line with the spec.

**Comment by ReinhardKeil** [comment]
I suggest we define a validation process for packs that are under development.  I believe it should be possible to use packchk in such scenarios and this is a much better tool for validation than adding messages to csolution.  For csolution we should have some tolerance implemented as this allows us to add new attributes over time.

**Comment by fred-r** [comment]
tolerance is fine for me, but why no warning ? 
The packchk can succeed because I have the latest version and the csolution may skip the condition because it is an older version.

**Comment by edriouk** [comment]
@fred-r , csolution does not perform schema check for the pdsc files. Therefore it cannot spot an unknown attribute. 
The condition evaluator has only very limited knowledge about attribute semantics. It matches the values in the supplied map  to the corresponding values in the expression (also a map). If an attribute is missing in one of those collections it is ignored (similar to SQL queries).  
Please note that in the case of an older tool, the supplied map won't contain an unknown attribute since the older tool won't be able to set its value and evaluation result will be correct.

**Comment by fred-r** [comment]
Skipping can be fine for me, I understand your point of view.
But my point is: why silently ?
I think we should get a warning because to me in this situation:

Pack A version 1 according to spec release 1:
condition is "Dname==xxx"
- packchk v1.0 succeeds
- csolution v1.0 computes the condition properly

Pack A version 2 according to spec release 2:
condition is "Dname==xxx && Dblabla==blabla"
- packchk v2.0 succeeds because it is aligned with the spec release 2
- csolution v1.0 computes the condition incorrectly because it skips the second part

**Comment by edriouk** [comment]
> But my point is: why silently ?
Because the evaluator knows neither  available attributes nor their possible values. Adding that information would require permanent update of all tools with each new attribute and wait for a new release of a tool.
It is still possible to ensure that even old tools evaluate new attribute correctly using deny expression
```
<condition id="newblabla"> 
<require Dname="xxx" Dblabla="blabla" /require>  
<deny Dname="xxx" Dblabla="" /deny>  
</condition>
```

**Comment by slhultgren** [comment]
For me this issue is more about the `who fills the value`.

In general, condition evaluation can just skip conditions if there are missing values, this is fine.

But I think the case is when:
* csolution sees that a condition is using a `Dblabla` in the condition
* csolution does not know how to populate a value for `Dblabla`, then this is different from a missing value, because the condition being evaluated is using "unknown potential features", which means that the csolution is potentially too old.
* A potential updated csolution could know how to populate the value for the `Dblabla`, perhaps using the exact same project files, but know more info from them.

Basically, if a new value is added to https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/pdsc_conditions_pg.html#element_accept, and csolution does not know which part of a project/pack populates the new property, it means that the new condition is "too new".

**Comment by edriouk** [comment]
1. csolution does not see a condition itself, but uses RTE Model library to evaluate expressions.
2. other tools that use RTE Model may be able to set an attribute value that is unknown to csolution (and vice versa)
3. some attributes are set transparently, device selection passes all device and processor attributes to condition evaluation. For this operation even outdated csolution works correctly by supplying correct value of an "unknown" attribute.

We can enhance RteConditionExpression::Validate() method with check for an unknown and deprecated attributes. 
However we need to pay keep in mind that:
a) the list of supported attributes should be created automatically during build process from the actual PACK.xsd. 
b) not all attributes listed in PACK.xsd might be already supported => we need a way to declare explicitly what is not supported 
c) validation should not affect condition evaluation and pack loading performance. I would suggest make validation only if requested, not by default

@fred-r , @slhultgren , would you volunteer to provide a PR with the suggested validation enhancements?

**Comment by jkrech** [comment]
There is obviously different ways to approach this compliance and implementation support. 
I think we should aim for the tools in the 'cmsis-toolbox' to support the features specified in their respective schema files. This will require to decouple the latest PACK.xsd from the specification to a version of the PACK.xsd that is supported by the tools.
Once we have the schema check incorporated in the `packchk` utility from the same cmsis-toolbox version should be quick, easy and consistent, especially for local packs under development.

**Comment by ReinhardKeil** [comment]
@fred-r is this still relevant?
